File: GreenSure/src/main/java/org/hartford/greensure/service/PolicyService.java
package org.hartford.greensure.service;

import org.hartford.greensure.dto.request.BuyPolicyRequest;
import org.hartford.greensure.dto.response.PolicyPlanResponse;
import org.hartford.greensure.dto.response.PolicyResponse;
import org.hartford.greensure.dto.response.UserPolicyResponse;
import org.hartford.greensure.entity.Policy;
import org.hartford.greensure.entity.PolicyPlan;
import org.hartford.greensure.entity.User;
import org.hartford.greensure.entity.UserPolicy;
import org.hartford.greensure.exception.BadRequestException;
import org.hartford.greensure.exception.ResourceNotFoundException;
import org.hartford.greensure.repository.PolicyPlanRepository;
import org.hartford.greensure.repository.PolicyRepository;
import org.hartford.greensure.repository.UserPolicyRepository;
import org.hartford.greensure.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class PolicyService {

    @Autowired private PolicyRepository policyRepository;
    @Autowired private PolicyPlanRepository policyPlanRepository;
    @Autowired private UserPolicyRepository userPolicyRepository;
    @Autowired private UserRepository userRepository;

    public List<PolicyResponse> getAvailablePolicies(Long userId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new ResourceNotFoundException("User not found"));

        String userType = user.getUserType().name();
        
        // Let's assume HOUSEHOLD sees null eligibility, MSME sees ALL
        List<Policy> policies;
        if ("HOUSEHOLD".equals(userType)) {
             policies = policyRepository.findByEligibilityOrEligibilityIsNull(null);
        } else {
             // MSME can see MSME and null
             policies = policyRepository.findAll();
        }

        return policies.stream().map(this::mapToResponse).collect(Collectors.toList());
    }

    @Transactional
    public UserPolicyResponse buyPolicy(Long userId, BuyPolicyRequest request) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new ResourceNotFoundException("User not found"));

        PolicyPlan plan = policyPlanRepository.findById(request.getPlanId())
                .orElseThrow(() -> new ResourceNotFoundException("Policy Plan not found"));

        UserPolicy userPolicy = UserPolicy.builder()
                .user(user)
                .plan(plan)
                .durationMonths(request.getDurationMonths())
                .finalPrice(request.getFinalPrice())
                .purchasedAt(LocalDateTime.now())
                .build();

        userPolicy = userPolicyRepository.save(userPolicy);
        return mapToUserPolicyResponse(userPolicy);
    }
    
    public List<UserPolicyResponse> getMyPolicies(Long userId) {
        return userPolicyRepository.findByUserUserIdOrderByPurchasedAtDesc(userId)
                .stream()
                .map(this::mapToUserPolicyResponse)
                .collect(Collectors.toList());
    }

    private PolicyResponse mapToResponse(Policy policy) {
        List<PolicyPlanResponse> planResponses = policy.getPlans().stream()
                .map(this::mapPlanToResponse)
                .collect(Collectors.toList());

        return PolicyResponse.builder()
                .policyType(policy.getPolicyType())
                .name(policy.getName())
                .icon(policy.getIcon())
                .description(policy.getDescription())
                .eligibility(policy.getEligibility())
                .plans(planResponses)
                .build();
    }

    private PolicyPlanResponse mapPlanToResponse(PolicyPlan plan) {
        return PolicyPlanResponse.builder()
                .planId(plan.getPlanId())
                .planName(plan.getPlanName())
                .coverageAmount(plan.getCoverageAmount())
                .basePremiumYearly(plan.getBasePremiumYearly())
                .features(plan.getFeatures())
                .build();
    }
    
    private UserPolicyResponse mapToUserPolicyResponse(UserPolicy up) {
        return UserPolicyResponse.builder()
                .id(up.getId())
                .userId(up.getUser().getUserId())
                .planId(up.getPlan().getPlanId())
                .policyType(up.getPlan().getPolicy().getPolicyType())
                .policyName(up.getPlan().getPolicy().getName())
                .planName(up.getPlan().getPlanName())
                .coverageAmount(up.getPlan().getCoverageAmount())
                .durationMonths(up.getDurationMonths())
                .finalPrice(up.getFinalPrice())
                .purchasedAt(up.getPurchasedAt())
                .build();
    }
}

File: greensure-frontend/src/app/core/services/policy.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { ApiResponse } from '../models/api-response';

export interface PolicyPlan {
    planId: number;
    planName: string;
    coverageAmount: number;
    basePremiumYearly: number;
    features: string[];
}

export interface Policy {
    policyType: string;
    name: string;
    icon: string;
    description: string;
    eligibility?: string;
    plans: PolicyPlan[];
}

export interface UserPolicy {
    id: number;
    userId: number;
    planId: number;
    policyType: string;
    policyName: string;
    planName: string;
    coverageAmount: number;
    durationMonths: number;
    finalPrice: number;
    purchasedAt: string;
}

@Injectable({
    providedIn: 'root'
})
export class PolicyService {
    private http = inject(HttpClient);
    private apiUrl = 'http://localhost:9090/api/policies';

    getPolicies(): Observable<ApiResponse<Policy[]>> {
        return this.http.get<ApiResponse<Policy[]>>(this.apiUrl);
    }

    buyPolicy(request: { planId: number; durationMonths: number; finalPrice: number }): Observable<ApiResponse<UserPolicy>> {
        return this.http.post<ApiResponse<UserPolicy>>(`${this.apiUrl}/buy`, request);
    }

    getMyPolicies(): Observable<ApiResponse<UserPolicy[]>> {
        return this.http.get<ApiResponse<UserPolicy[]>>(`${this.apiUrl}/my-policies`);
    }
}
